File: aoc2022/day03.py
from .utils import Answer

def priority(c):
    if 'a' <= c <= 'z':
        return ord(c) - 96
    return ord(c) - 38

def solve(input: str):
    lines = input.split()
    p1, p2 = 0, 0
    for line in lines:
        d = len(line) // 2
        left, right = line[:d], line[d:]
        right = line[d:]
        for common in set(left) & set(right):
            p1 += priority(common)

    for i in range(0, len(lines), 3):
        sets = map(set, lines[i:i+3])
        badge = set.intersection(*sets)
        p2 += priority(badge.pop())

    return Answer(p1, p2)

File: src/day03.nim
import common
import sets
import sequtils

proc priority(c: char): int =
  if 'a' <= c and c <= 'z':
    return ord(c)-96
  else:
    return ord(c)-38

proc solve*(input: string): Answer =
  let lines = toSeq(lines input)
  for line in lines:
    let d = line.len div 2
    let (left, right) = (line[0 .. d-1], line[d .. line.high])
    let common = left.toHashSet * right.toHashSet
    for c in common:
      result.part1 += c.priority

  for i in countup(0, lines.high, 3):
    let a = lines[i].toHashSet
    let b = lines[i+1].toHashSet
    let c = lines[i+2].toHashSet
    var badge = a * b * c
    result.part2 += badge.pop.priority
